### ISSUE TITLE
[JENKINS-21771] Apparent increase in thread usage for slave nodes

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/ssh-agents-plugin#763]
<p>After upgrading to Jenkins v1.549, ssh-slaves v1.6, git plugin v2.0.1, git-client plugin v1.6.2, we began observing many errors and job failures when running builds on our slave nodes. The errors were typically of the form:</p>

<pre>

 Caused by: java.lang.OutOfMemoryError: unable to create <span class="code-keyword">new</span> <span class="code-keyword">native</span> thread</pre>
<p>In general, we saw these errors mainly during two stages of the build process: during SCM step (git checkout), and during transfer of a file specified as a job parameter to the slave node. Some examples of these errors are below; however, it is worth noting that we did observe errors at other phases of the build as well (i.e. was not limited to these two phases).</p>

<pre>

00:00:45.338 ERROR: Workspace has a .git repository, but it appears to be corrupt.
00:00:45.972 FATAL: java.io.IOException: Remote call on bob0024 failed
00:00:45.972 hudson.remoting.RemotingSystemException: java.io.IOException: Remote call on bob0024 failed
00:00:45.973 at hudson.remoting.RemoteInvocationHandler.invoke(RemoteInvocationHandler.java:183)
00:00:45.973 at $Proxy65.hasGitRepo(Unknown Source)
00:00:45.973 at org.jenkinsci.plugins.gitclient.RemoteGitImpl.hasGitRepo(RemoteGitImpl.java:250)
00:00:45.973 at hudson.plugins.git.GitSCM.retrieveChanges(GitSCM.java:824)
00:00:45.973 at hudson.plugins.git.GitSCM.checkout(GitSCM.java:872)
00:00:45.973 at org.jenkinsci.plugins.multiplescms.MultiSCM.checkout(MultiSCM.java:118)
00:00:45.973 at hudson.model.AbstractProject.checkout(AbstractProject.java:1411)
00:00:45.973 at hudson.model.AbstractBuild$AbstractBuildExecution.defaultCheckout(AbstractBuild.java:651)
00:00:45.973 at jenkins.scm.SCMCheckoutStrategy.checkout(SCMCheckoutStrategy.java:88)
00:00:45.973 at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:560)
00:00:45.973 at hudson.model.Run.execute(Run.java:1670)
00:00:45.973 at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:46)
00:00:45.973 at hudson.model.ResourceController.execute(ResourceController.java:88)
00:00:45.973 at hudson.model.Executor.run(Executor.java:231)
00:00:45.973 Caused by: java.io.IOException: Remote call on bob0024 failed
00:00:45.973 at hudson.remoting.Channel.call(Channel.java:731)
00:00:45.973 at hudson.remoting.RemoteInvocationHandler.invoke(RemoteInvocationHandler.java:167)
00:00:45.973 ... 13 more
00:00:45.973 Caused by: java.lang.OutOfMemoryError: unable to create <span class="code-keyword">new</span> <span class="code-keyword">native</span> thread
00:00:45.973 at java.lang.<span class="code-object">Thread</span>.start0(Native Method)
00:00:45.973 at java.lang.<span class="code-object">Thread</span>.start(<span class="code-object">Thread</span>.java:597)
00:00:45.973 at hudson.remoting.AtmostOneThreadExecutor.execute(AtmostOneThreadExecutor.java:88)
00:00:45.973 at java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:78)
00:00:45.973 at hudson.remoting.JarCacheSupport.resolve(JarCacheSupport.java:59)
00:00:45.973 at hudson.remoting.ResourceImageBoth.initiateJarRetrieval(ResourceImageBoth.java:40)
00:00:45.973 at hudson.remoting.ResourceImageBoth.resolve(ResourceImageBoth.java:22)
00:00:45.973 at hudson.remoting.RemoteClassLoader.findClass(RemoteClassLoader.java:233)
00:00:45.973 at java.lang.<span class="code-object">ClassLoader</span>.loadClass(<span class="code-object">ClassLoader</span>.java:307)
00:00:45.973 at java.lang.<span class="code-object">ClassLoader</span>.loadClass(<span class="code-object">ClassLoader</span>.java:248)
00:00:45.973 at hudson.util.StreamTaskListener._error(StreamTaskListener.java:132)
00:00:45.973 at hudson.util.StreamTaskListener.error(StreamTaskListener.java:141)
00:00:45.973 at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.hasGitRepo(CliGitAPIImpl.java:126)
00:00:45.973 at hudson.plugins.git.GitAPI.hasGitRepo(GitAPI.java:186)
00:00:45.973 at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
00:00:45.973 at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:39)
00:00:45.973 at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:25)
00:00:45.973 at java.lang.reflect.Method.invoke(Method.java:597)
00:00:45.973 at hudson.remoting.RemoteInvocationHandler$RPCRequest.perform(RemoteInvocationHandler.java:299)
00:00:45.973 at hudson.remoting.RemoteInvocationHandler$RPCRequest.call(RemoteInvocationHandler.java:280)
00:00:45.973 at hudson.remoting.RemoteInvocationHandler$RPCRequest.call(RemoteInvocationHandler.java:239)
00:00:45.973 at hudson.remoting.UserRequest.perform(UserRequest.java:118)
00:00:45.973 at hudson.remoting.UserRequest.perform(UserRequest.java:48)
00:00:45.973 at hudson.remoting.Request$2.run(Request.java:328)
00:00:45.973 at hudson.remoting.InterceptingExecutorService$1.call(InterceptingExecutorService.java:72)
00:00:45.973 at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:303)
00:00:45.973 at java.util.concurrent.FutureTask.run(FutureTask.java:138)
00:00:45.973 at java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:886)
00:00:45.973 at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:908)
00:00:45.973 at java.lang.<span class="code-object">Thread</span>.run(<span class="code-object">Thread</span>.java:619)
</pre>

<p>When such an error was observed in the job console output (as above), the following was seen in the slave log:</p>

<pre>

SEVERE: Failed to execute command UserRequest:hudson.node_monitors.ResponseTimeMonitor$Step1@​34c82420 (channel channel)
java.lang.OutOfMemoryError: unable to create <span class="code-keyword">new</span> <span class="code-keyword">native</span> thread
at java.lang.<span class="code-object">Thread</span>.start0(Native Method)
at java.lang.<span class="code-object">Thread</span>.start(<span class="code-object">Thread</span>.java:597)
at java.util.concurrent.ThreadPoolExecutor.addIfUnderMaximumPoolSize(ThreadPoolExecutor.java:727)
at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:657)
at java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:92)
at hudson.remoting.DelegatingExecutorService.submit(DelegatingExecutorService.java:42)
at hudson.remoting.InterceptingExecutorService.submit(InterceptingExecutorService.java:46)
at hudson.remoting.InterceptingExecutorService.submit(InterceptingExecutorService.java:41)
at hudson.remoting.Request.execute(Request.java:307)
at hudson.remoting.Channel$2.handle(Channel.java:461)
at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:60)
Feb 10, 2014 5:24:41 PM hudson.remoting.Channel$2 handle
SEVERE: This command is created here
Command UserRequest:hudson.node_monitors.ResponseTimeMonitor$Step1@​34c82420 created at
at hudson.remoting.Command.<init>(Command.java:67)
at hudson.remoting.Command.<init>(Command.java:50)
at hudson.remoting.Request.<init>(Request.java:98)
at hudson.remoting.UserRequest.<init>(UserRequest.java:59)
at hudson.remoting.Channel.callAsync(Channel.java:749)
at hudson.node_monitors.AbstractAsyncNodeMonitorDescriptor.monitor(AbstractAsyncNodeMonitorDescriptor.java:76)
at hudson.node_monitors.ResponseTimeMonitor$1.monitor(ResponseTimeMonitor.java:57)
at hudson.node_monitors.AbstractNodeMonitorDescriptor$Record.run(AbstractNodeMonitorDescriptor.java:280)
Feb 10, 2014 5:24:41 PM hudson.remoting.Channel$2 handle
SEVERE: Failed to execute command UserRequest:hudson.node_monitors.SwapSpaceMonitor$MonitorTask@​438a24d1 (channel channel)
java.lang.OutOfMemoryError: unable to create <span class="code-keyword">new</span> <span class="code-keyword">native</span> thread
at java.lang.<span class="code-object">Thread</span>.start0(Native Method)
at java.lang.<span class="code-object">Thread</span>.start(<span class="code-object">Thread</span>.java:597)
at java.util.concurrent.ThreadPoolExecutor.addIfUnderMaximumPoolSize(ThreadPoolExecutor.java:727)
at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:657)
at java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:92)
at hudson.remoting.DelegatingExecutorService.submit(DelegatingExecutorService.java:42)
at hudson.remoting.InterceptingExecutorService.submit(InterceptingExecutorService.java:46)
at hudson.remoting.InterceptingExecutorService.submit(InterceptingExecutorService.java:41)
at hudson.remoting.Request.execute(Request.java:307)
at hudson.remoting.Channel$2.handle(Channel.java:461)
at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:60)
Feb 10, 2014 5:24:41 PM hudson.remoting.Channel$2 handle
SEVERE: This command is created here
Command UserRequest:hudson.node_monitors.SwapSpaceMonitor$MonitorTask@​438a24d1 created at
at hudson.remoting.Command.<init>(Command.java:67)
at hudson.remoting.Command.<init>(Command.java:50)
at hudson.remoting.Request.<init>(Request.java:98)
at hudson.remoting.UserRequest.<init>(UserRequest.java:59)
at hudson.remoting.Channel.callAsync(Channel.java:749)
at hudson.node_monitors.AbstractAsyncNodeMonitorDescriptor.monitor(AbstractAsyncNodeMonitorDescriptor.java:76)
at hudson.node_monitors.AbstractNodeMonitorDescriptor$Record.run(AbstractNodeMonitorDescriptor.java:280)
Feb 10, 2014 5:24:41 PM hudson.remoting.Channel$2 handle
SEVERE: Failed to execute command Pipe.Chunk(4,8192) (channel channel)
java.lang.OutOfMemoryError: unable to create <span class="code-keyword">new</span> <span class="code-keyword">native</span> thread
at java.lang.<span class="code-object">Thread</span>.start0(Native Method)
at java.lang.<span class="code-object">Thread</span>.start(<span class="code-object">Thread</span>.java:597)
at java.util.concurrent.ThreadPoolExecutor.addIfUnderMaximumPoolSize(ThreadPoolExecutor.java:727)
at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:657)
at java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:92)
at hudson.remoting.DelegatingExecutorService.submit(DelegatingExecutorService.java:42)
at hudson.remoting.InterceptingExecutorService.submit(InterceptingExecutorService.java:46)
at hudson.remoting.InterceptingExecutorService.submit(InterceptingExecutorService.java:41)
at hudson.remoting.SingleLaneExecutorService.execute(SingleLaneExecutorService.java:104)
at java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:78)
at hudson.remoting.PipeWriter.submit(PipeWriter.java:152)
at hudson.remoting.ProxyOutputStream$Chunk.execute(ProxyOutputStream.java:247)
at hudson.remoting.Channel$2.handle(Channel.java:461)
at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:60)
Feb 10, 2014 5:24:41 PM hudson.remoting.Channel$2 handle
SEVERE: This command is created here
Feb 10, 2014 5:24:41 PM hudson.remoting.Channel$2 handle
SEVERE: Failed to execute command UserRequest:hudson.FilePath$9@​480e8c6 (channel channel)
java.lang.OutOfMemoryError: unable to create <span class="code-keyword">new</span> <span class="code-keyword">native</span> thread
at java.lang.<span class="code-object">Thread</span>.start0(Native Method)
at java.lang.<span class="code-object">Thread</span>.start(<span class="code-object">Thread</span>.java:597)
at java.util.concurrent.ThreadPoolExecutor.addIfUnderMaximumPoolSize(ThreadPoolExecutor.java:727)
at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:657)
at java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:92)
at hudson.remoting.DelegatingExecutorService.submit(DelegatingExecutorService.java:42)
at hudson.remoting.InterceptingExecutorService.submit(InterceptingExecutorService.java:46)
at hudson.remoting.InterceptingExecutorService.submit(InterceptingExecutorService.java:41)
at hudson.remoting.Request.execute(Request.java:307)
at hudson.remoting.Channel$2.handle(Channel.java:461)
at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:60)
SEVERE: This command is created here
Command UserRequest:hudson.node_monitors.ResponseTimeMonitor$Step1@​76ff540f created at
at hudson.remoting.Command.<init>(Command.java:67)
at hudson.remoting.Command.<init>(Command.java:50)
at hudson.remoting.Request.<init>(Request.java:98)
at hudson.remoting.UserRequest.<init>(UserRequest.java:59)
at hudson.remoting.Channel.callAsync(Channel.java:749)
at hudson.node_monitors.AbstractAsyncNodeMonitorDescriptor.monitor(AbstractAsyncNodeMonitorDescriptor.java:76)
at hudson.node_monitors.ResponseTimeMonitor$1.monitor(ResponseTimeMonitor.java:57)
at hudson.node_monitors.AbstractNodeMonitorDescriptor$Record.run(AbstractNodeMonitorDescriptor.java:280)
Feb 10, 2014 5:25:17 PM hudson.remoting.Channel$2 handle
SEVERE: Failed to execute command UserRequest:hudson.FilePath$9@​4a50a2df (channel channel)
java.lang.OutOfMemoryError: unable to create <span class="code-keyword">new</span> <span class="code-keyword">native</span> thread
at java.lang.<span class="code-object">Thread</span>.start0(Native Method)
at java.lang.<span class="code-object">Thread</span>.start(<span class="code-object">Thread</span>.java:597)
at java.util.concurrent.ThreadPoolExecutor.addIfUnderMaximumPoolSize(ThreadPoolExecutor.java:727)
at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:657)
at java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:92)
at hudson.remoting.DelegatingExecutorService.submit(DelegatingExecutorService.java:42)
at hudson.remoting.InterceptingExecutorService.submit(InterceptingExecutorService.java:46)
at hudson.remoting.InterceptingExecutorService.submit(InterceptingExecutorService.java:41)
at hudson.remoting.Request.execute(Request.java:307)
at hudson.remoting.Channel$2.handle(Channel.java:461)
at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:60)
Feb 10, 2014 5:25:17 PM hudson.remoting.Channel$2 handle
SEVERE: This command is created here
Command UserRequest:hudson.FilePath$9@​4a50a2df created at
at hudson.remoting.Command.<init>(Command.java:67)
at hudson.remoting.Command.<init>(Command.java:50)
at hudson.remoting.Request.<init>(Request.java:98)
at hudson.remoting.UserRequest.<init>(UserRequest.java:59)
at hudson.remoting.Channel.callAsync(Channel.java:749)
at hudson.node_monitors.AbstractAsyncNodeMonitorDescriptor.monitor(AbstractAsyncNodeMonitorDescriptor.java:76)
at hudson.node_monitors.AbstractNodeMonitorDescriptor$Record.run(AbstractNodeMonitorDescriptor.java:280)
Feb 10, 2014 5:25:18 PM hudson.remoting.Channel$2 handle
SEVERE: Failed to execute command UserRequest:hudson.FilePath$9@​77d5fff1 (channel channel)
java.lang.OutOfMemoryError: unable to create <span class="code-keyword">new</span> <span class="code-keyword">native</span> thread
at java.lang.<span class="code-object">Thread</span>.start0(Native Method)
at java.lang.<span class="code-object">Thread</span>.start(<span class="code-object">Thread</span>.java:597)
at java.util.concurrent.ThreadPoolExecutor.addIfUnderMaximumPoolSize(ThreadPoolExecutor.java:727)
at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:657)
at java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:92)
at hudson.remoting.DelegatingExecutorService.submit(DelegatingExecutorService.java:42)
at hudson.remoting.InterceptingExecutorService.submit(InterceptingExecutorService.java:46)
at hudson.remoting.InterceptingExecutorService.submit(InterceptingExecutorService.java:41)
at hudson.remoting.Request.execute(Request.java:307)
at hudson.remoting.Channel$2.handle(Channel.java:461)
at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:60)
Feb 10, 2014 5:25:18 PM hudson.remoting.Channel$2 handle
SEVERE: This command is created here
Command UserRequest:hudson.FilePath$9@​77d5fff1 created at
at hudson.remoting.Command.<init>(Command.java:67)
at hudson.remoting.Command.<init>(Command.java:50)
at hudson.remoting.Request.<init>(Request.java:98)
at hudson.remoting.UserRequest.<init>(UserRequest.java:59)
at hudson.remoting.Channel.callAsync(Channel.java:749)
at hudson.node_monitors.AbstractAsyncNodeMonitorDescriptor.monitor(AbstractAsyncNodeMonitorDescriptor.java:76)
at hudson.node_monitors.AbstractNodeMonitorDescriptor$Record.run(AbstractNodeMonitorDescriptor.java:280)
Feb 10, 2014 5:25:18 PM hudson.remoting.Channel$2 handle
SEVERE: Failed to execute command UserRequest:hudson.FilePath$9@​3c51eb73 (channel channel)
java.lang.OutOfMemoryError: unable to create <span class="code-keyword">new</span> <span class="code-keyword">native</span> thread
at java.lang.<span class="code-object">Thread</span>.start0(Native Method)
at java.lang.<span class="code-object">Thread</span>.start(<span class="code-object">Thread</span>.java:597)
at java.util.concurrent.ThreadPoolExecutor.addIfUnderMaximumPoolSize(ThreadPoolExecutor.java:727)
at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:657)
at java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:92)
at hudson.remoting.DelegatingExecutorService.submit(DelegatingExecutorService.java:42)
at hudson.remoting.InterceptingExecutorService.submit(InterceptingExecutorService.java:46)
at hudson.remoting.InterceptingExecutorService.submit(InterceptingExecutorService.java:41)
at hudson.remoting.Request.execute(Request.java:307)
at hudson.remoting.Channel$2.handle(Channel.java:461)
at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:60)
Feb 10, 2014 5:25:18 PM hudson.remoting.Channel$2 handle
SEVERE: This command is created here
Command UserRequest:hudson.FilePath$9@​3c51eb73 created at
at hudson.remoting.Command.<init>(Command.java:67)
at hudson.remoting.Command.<init>(Command.java:50)
at hudson.remoting.Request.<init>(Request.java:98)
at hudson.remoting.UserRequest.<init>(UserRequest.java:59)
at hudson.remoting.Channel.callAsync(Channel.java:749)
at hudson.node_monitors.AbstractAsyncNodeMonitorDescriptor.monitor(AbstractAsyncNodeMonitorDescriptor.java:76)
at hudson.node_monitors.AbstractNodeMonitorDescriptor$Record.run(AbstractNodeMonitorDescriptor.java:280)
2014-02-10 18:53:15
Full thread dump Java HotSpot(TM) 64-Bit Server VM (16.3-b01 mixed mode):

<span class="code-quote">"pool-1-thread-281"</span> prio=10 tid=0x00007fa9bc001000 nid=0x3150 waiting on condition [0x00007faa15bdc000]
java.lang.<span class="code-object">Thread</span>.State: TIMED_WAITING (parking)
at sun.misc.Unsafe.park(Native Method)
 - parking to wait <span class="code-keyword">for</span> <0x00007facf8c32d60> (a java.util.concurrent.SynchronousQueue$TransferStack)
at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:198)
at java.util.concurrent.SynchronousQueue$TransferStack.awaitFulfill(SynchronousQueue.java:424)
at java.util.concurrent.SynchronousQueue$TransferStack.transfer(SynchronousQueue.java:323)
at java.util.concurrent.SynchronousQueue.poll(SynchronousQueue.java:874)
at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:945)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:907)
at java.lang.<span class="code-object">Thread</span>.run(<span class="code-object">Thread</span>.java:619)

<span class="code-quote">"pool-1-thread-279 <span class="code-keyword">for</span> channel"</span> prio=10 tid=0x00007fa9bc00e800 nid=0x2613 waiting on condition [0x00007faa157d8000]
java.lang.<span class="code-object">Thread</span>.State: WAITING (parking)
at sun.misc.Unsafe.park(Native Method)
 - parking to wait <span class="code-keyword">for</span> <0x00007face7a65b58> (a java.util.concurrent.FutureTask$Sync)
at java.util.concurrent.locks.LockSupport.park(LockSupport.java:158)
at java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(AbstractQueuedSynchronizer.java:747)
at java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireSharedInterruptibly(AbstractQueuedSynchronizer.java:905)
at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireSharedInterruptibly(AbstractQueuedSynchronizer.java:1217)
at java.util.concurrent.FutureTask$Sync.innerGet(FutureTask.java:218)
at java.util.concurrent.FutureTask.get(FutureTask.java:83)
at hudson.remoting.Request$2.run(Request.java:326)
at hudson.remoting.InterceptingExecutorService$1.call(InterceptingExecutorService.java:72)
at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:303)
at java.util.concurrent.FutureTask.run(FutureTask.java:138)
at java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:886)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:908)
at java.lang.<span class="code-object">Thread</span>.run(<span class="code-object">Thread</span>.java:619)

<span class="code-quote">"Ping thread <span class="code-keyword">for</span> channel hudson.remoting.Channel@​37bd2664:channel"</span> daemon prio=10 tid=0x00007fa9b4020800 nid=0x4158 waiting on condition [0x00007faa16dec000]
java.lang.<span class="code-object">Thread</span>.State: TIMED_WAITING (sleeping)
at java.lang.<span class="code-object">Thread</span>.sleep(Native Method)
at hudson.remoting.PingThread.run(PingThread.java:86)

<span class="code-quote">"<span class="code-object">Thread</span>-1 / waiting <span class="code-keyword">for</span> hudson.remoting.Channel@​37bd2664:channel"</span> daemon prio=10 tid=0x00007fa9b4003000 nid=0x4156 in <span class="code-object">Object</span>.wait() [0x00007faa16fee000]
java.lang.<span class="code-object">Thread</span>.State: TIMED_WAITING (on object monitor)
at java.lang.<span class="code-object">Object</span>.wait(Native Method)
 - waiting on <0x00007facf7f88240> (a hudson.remoting.RemoteInvocationHandler$RPCRequest)
at hudson.remoting.Request.call(Request.java:146)
 - locked <0x00007facf7f88240> (a hudson.remoting.RemoteInvocationHandler$RPCRequest)
at hudson.remoting.RemoteInvocationHandler.invoke(RemoteInvocationHandler.java:170)
at $Proxy4.writeJarTo(Unknown Source)
at hudson.remoting.FileSystemJarCache.retrieve(FileSystemJarCache.java:66)
at hudson.remoting.JarCacheSupport$1.run(JarCacheSupport.java:62)
at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:441)
at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:303)
at java.util.concurrent.FutureTask.run(FutureTask.java:138)
at hudson.remoting.AtmostOneThreadExecutor$Worker.run(AtmostOneThreadExecutor.java:104)
at java.lang.<span class="code-object">Thread</span>.run(<span class="code-object">Thread</span>.java:619)

<span class="code-quote">"Ping thread <span class="code-keyword">for</span> channel hudson.remoting.Channel@​37bd2664:channel"</span> daemon prio=10 tid=0x00007fae381ae800 nid=0x4154 waiting on condition [0x00007faa171f0000]
java.lang.<span class="code-object">Thread</span>.State: TIMED_WAITING (sleeping)
at java.lang.<span class="code-object">Thread</span>.sleep(Native Method)
at hudson.remoting.PingThread.run(PingThread.java:86)

<span class="code-quote">"Channel reader thread: channel"</span> prio=10 tid=0x00007fae381bf800 nid=0x4153 runnable [0x00007faa172f1000]
java.lang.<span class="code-object">Thread</span>.State: RUNNABLE
at java.io.FileInputStream.readBytes(Native Method)
at java.io.FileInputStream.read(FileInputStream.java:199)
at java.io.BufferedInputStream.fill(BufferedInputStream.java:218)
at java.io.BufferedInputStream.read(BufferedInputStream.java:237)
 - locked <0x00007facf7f921f8> (a java.io.BufferedInputStream)
at hudson.remoting.FlightRecorderInputStream.read(FlightRecorderInputStream.java:77)
at java.io.ObjectInputStream$PeekInputStream.peek(ObjectInputStream.java:2249)
at java.io.ObjectInputStream$BlockDataInputStream.peek(ObjectInputStream.java:2542)
at java.io.ObjectInputStream$BlockDataInputStream.peekByte(ObjectInputStream.java:2552)
at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1297)
at java.io.ObjectInputStream.readObject(ObjectInputStream.java:351)
at hudson.remoting.Command.readFrom(Command.java:92)
at hudson.remoting.ClassicCommandTransport.read(ClassicCommandTransport.java:71)
at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:48)

<span class="code-quote">"Low Memory Detector"</span> daemon prio=10 tid=0x00007fae380d9800 nid=0x4151 runnable [0x0000000000000000]
java.lang.<span class="code-object">Thread</span>.State: RUNNABLE

<span class="code-quote">"CompilerThread1"</span> daemon prio=10 tid=0x00007fae380d7000 nid=0x4150 waiting on condition [0x0000000000000000]
java.lang.<span class="code-object">Thread</span>.State: RUNNABLE

<span class="code-quote">"CompilerThread0"</span> daemon prio=10 tid=0x00007fae380d4800 nid=0x414f waiting on condition [0x0000000000000000]
java.lang.<span class="code-object">Thread</span>.State: RUNNABLE

<span class="code-quote">"Signal Dispatcher"</span> daemon prio=10 tid=0x00007fae380d2800 nid=0x414e waiting on condition [0x0000000000000000]
java.lang.<span class="code-object">Thread</span>.State: RUNNABLE

<span class="code-quote">"Finalizer"</span> daemon prio=10 tid=0x00007fae380b4800 nid=0x414d in <span class="code-object">Object</span>.wait() [0x00007faa178f7000]
java.lang.<span class="code-object">Thread</span>.State: WAITING (on object monitor)
at java.lang.<span class="code-object">Object</span>.wait(Native Method)
 - waiting on <0x00007facf7f927c0> (a java.lang.ref.ReferenceQueue$Lock)
at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:118)
 - locked <0x00007facf7f927c0> (a java.lang.ref.ReferenceQueue$Lock)
at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:134)
at java.lang.ref.Finalizer$FinalizerThread.run(Finalizer.java:159)

<span class="code-quote">"Reference Handler"</span> daemon prio=10 tid=0x00007fae380b2800 nid=0x414c in <span class="code-object">Object</span>.wait() [0x00007faa179f8000]
java.lang.<span class="code-object">Thread</span>.State: WAITING (on object monitor)
at java.lang.<span class="code-object">Object</span>.wait(Native Method)
 - waiting on <0x00007facf7f8b760> (a java.lang.ref.Reference$Lock)
at java.lang.<span class="code-object">Object</span>.wait(<span class="code-object">Object</span>.java:485)
at java.lang.ref.Reference$ReferenceHandler.run(Reference.java:116)
 - locked <0x00007facf7f8b760> (a java.lang.ref.Reference$Lock)

<span class="code-quote">"main"</span> prio=10 tid=0x00007fae38006000 nid=0x4133 in <span class="code-object">Object</span>.wait() [0x00007fae3d158000]
java.lang.<span class="code-object">Thread</span>.State: WAITING (on object monitor)
at java.lang.<span class="code-object">Object</span>.wait(Native Method)
 - waiting on <0x00007facf7f882a8> (a hudson.remoting.Channel)
at java.lang.<span class="code-object">Object</span>.wait(<span class="code-object">Object</span>.java:485)
at hudson.remoting.Channel.join(Channel.java:850)
 - locked <0x00007facf7f882a8> (a hudson.remoting.Channel)
at hudson.remoting.Launcher.main(Launcher.java:523)
at hudson.remoting.Launcher.runWithStdinStdout(Launcher.java:457)
at hudson.remoting.Launcher.run(Launcher.java:238)
at hudson.remoting.Launcher.main(Launcher.java:192)
</pre>

<p>These errors were not dependably reproducible, and were nearly impossible to duplicate in isolation (i.e. could not reproduce when just running jobs on a single slave node).</p>

<p>After some time debugging and ruling out some things, we determined that perhaps the user nproc limit (ulimit -u) might be too low, based on the errors that w were seeing and what we were able to rule out after debugging. We increased this limit on our slave nodes and...things went back to normal.</p>

<p>None of our build and test jobs changed across the update, meaning that any changes we observed were introduced as a result of the upgrade. My question is then: what has changed recently with slave management that has increased the number of per node threads required? We don't yet have solid watermark data on how high it was actually going, but it was high enough to hit the ceiling for system default (1024).</p>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/greg.png" title="greg's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=greg">greg</a>, imported from: <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-21771" target="_blank">Apparent increase in thread usage for slave nodes</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/ifernandezcalvo.png" title="ifernandezcalvo's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=ifernandezcalvo">ifernandezcalvo</a>
<li><b>status</b>: Closed
<li><b>priority</b>: Major
<li><b>component(s)</b>: ssh-slaves-plugin
<li><b>label(s)</b>: jenkins, performance, slave
<li><b>resolution</b>: Won't Do
<li><b>resolved</b>: 2018-07-20T18:49:09+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 2
<li><b>imported</b>: 20251212-090250
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>After upgrading to Jenkins v1.549, ssh-slaves v1.6, git plugin v2.0.1, git-client plugin v1.6.2, we began observing many errors and job failures when running builds on our slave nodes. The errors were typically of the form:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">
 Caused by: java.lang.OutOfMemoryError: unable to create <span class="code-keyword">new</span> <span class="code-keyword">native</span> thread</pre>
</div></div>
<p>In general, we saw these errors mainly during two stages of the build process: during SCM step (git checkout), and during transfer of a file specified as a job parameter to the slave node. Some examples of these errors are below; however, it is worth noting that we did observe errors at other phases of the build as well (i.e. was not limited to these two phases).</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">
00:00:45.338 ERROR: Workspace has a .git repository, but it appears to be corrupt.
00:00:45.972 FATAL: java.io.IOException: Remote call on bob0024 failed
00:00:45.972 hudson.remoting.RemotingSystemException: java.io.IOException: Remote call on bob0024 failed
00:00:45.973 at hudson.remoting.RemoteInvocationHandler.invoke(RemoteInvocationHandler.java:183)
00:00:45.973 at $Proxy65.hasGitRepo(Unknown Source)
00:00:45.973 at org.jenkinsci.plugins.gitclient.RemoteGitImpl.hasGitRepo(RemoteGitImpl.java:250)
00:00:45.973 at hudson.plugins.git.GitSCM.retrieveChanges(GitSCM.java:824)
00:00:45.973 at hudson.plugins.git.GitSCM.checkout(GitSCM.java:872)
00:00:45.973 at org.jenkinsci.plugins.multiplescms.MultiSCM.checkout(MultiSCM.java:118)
00:00:45.973 at hudson.model.AbstractProject.checkout(AbstractProject.java:1411)
00:00:45.973 at hudson.model.AbstractBuild$AbstractBuildExecution.defaultCheckout(AbstractBuild.java:651)
00:00:45.973 at jenkins.scm.SCMCheckoutStrategy.checkout(SCMCheckoutStrategy.java:88)
00:00:45.973 at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:560)
00:00:45.973 at hudson.model.Run.execute(Run.java:1670)
00:00:45.973 at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:46)
00:00:45.973 at hudson.model.ResourceController.execute(ResourceController.java:88)
00:00:45.973 at hudson.model.Executor.run(Executor.java:231)
00:00:45.973 Caused by: java.io.IOException: Remote call on bob0024 failed
00:00:45.973 at hudson.remoting.Channel.call(Channel.java:731)
00:00:45.973 at hudson.remoting.RemoteInvocationHandler.invoke(RemoteInvocationHandler.java:167)
00:00:45.973 ... 13 more
00:00:45.973 Caused by: java.lang.OutOfMemoryError: unable to create <span class="code-keyword">new</span> <span class="code-keyword">native</span> thread
00:00:45.973 at java.lang.<span class="code-object">Thread</span>.start0(Native Method)
00:00:45.973 at java.lang.<span class="code-object">Thread</span>.start(<span class="code-object">Thread</span>.java:597)
00:00:45.973 at hudson.remoting.AtmostOneThreadExecutor.execute(AtmostOneThreadExecutor.java:88)
00:00:45.973 at java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:78)
00:00:45.973 at hudson.remoting.JarCacheSupport.resolve(JarCacheSupport.java:59)
00:00:45.973 at hudson.remoting.ResourceImageBoth.initiateJarRetrieval(ResourceImageBoth.java:40)
00:00:45.973 at hudson.remoting.ResourceImageBoth.resolve(ResourceImageBoth.java:22)
00:00:45.973 at hudson.remoting.RemoteClassLoader.findClass(RemoteClassLoader.java:233)
00:00:45.973 at java.lang.<span class="code-object">ClassLoader</span>.loadClass(<span class="code-object">ClassLoader</span>.java:307)
00:00:45.973 at java.lang.<span class="code-object">ClassLoader</span>.loadClass(<span class="code-object">ClassLoader</span>.java:248)
00:00:45.973 at hudson.util.StreamTaskListener._error(StreamTaskListener.java:132)
00:00:45.973 at hudson.util.StreamTaskListener.error(StreamTaskListener.java:141)
00:00:45.973 at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.hasGitRepo(CliGitAPIImpl.java:126)
00:00:45.973 at hudson.plugins.git.GitAPI.hasGitRepo(GitAPI.java:186)
00:00:45.973 at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
00:00:45.973 at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:39)
00:00:45.973 at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:25)
00:00:45.973 at java.lang.reflect.Method.invoke(Method.java:597)
00:00:45.973 at hudson.remoting.RemoteInvocationHandler$RPCRequest.perform(RemoteInvocationHandler.java:299)
00:00:45.973 at hudson.remoting.RemoteInvocationHandler$RPCRequest.call(RemoteInvocationHandler.java:280)
00:00:45.973 at hudson.remoting.RemoteInvocationHandler$RPCRequest.call(RemoteInvocationHandler.java:239)
00:00:45.973 at hudson.remoting.UserRequest.perform(UserRequest.java:118)
00:00:45.973 at hudson.remoting.UserRequest.perform(UserRequest.java:48)
00:00:45.973 at hudson.remoting.Request$2.run(Request.java:328)
00:00:45.973 at hudson.remoting.InterceptingExecutorService$1.call(InterceptingExecutorService.java:72)
00:00:45.973 at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:303)
00:00:45.973 at java.util.concurrent.FutureTask.run(FutureTask.java:138)
00:00:45.973 at java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:886)
00:00:45.973 at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:908)
00:00:45.973 at java.lang.<span class="code-object">Thread</span>.run(<span class="code-object">Thread</span>.java:619)
</pre>
</div></div>

<p>When such an error was observed in the job console output (as above), the following was seen in the slave log:</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">
SEVERE: Failed to execute command UserRequest:hudson.node_monitors.ResponseTimeMonitor$Step1@34c82420 (channel channel)
java.lang.OutOfMemoryError: unable to create <span class="code-keyword">new</span> <span class="code-keyword">native</span> thread
at java.lang.<span class="code-object">Thread</span>.start0(Native Method)
at java.lang.<span class="code-object">Thread</span>.start(<span class="code-object">Thread</span>.java:597)
at java.util.concurrent.ThreadPoolExecutor.addIfUnderMaximumPoolSize(ThreadPoolExecutor.java:727)
at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:657)
at java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:92)
at hudson.remoting.DelegatingExecutorService.submit(DelegatingExecutorService.java:42)
at hudson.remoting.InterceptingExecutorService.submit(InterceptingExecutorService.java:46)
at hudson.remoting.InterceptingExecutorService.submit(InterceptingExecutorService.java:41)
at hudson.remoting.Request.execute(Request.java:307)
at hudson.remoting.Channel$2.handle(Channel.java:461)
at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:60)
Feb 10, 2014 5:24:41 PM hudson.remoting.Channel$2 handle
SEVERE: This command is created here
Command UserRequest:hudson.node_monitors.ResponseTimeMonitor$Step1@34c82420 created at
at hudson.remoting.Command.&lt;init&gt;(Command.java:67)
at hudson.remoting.Command.&lt;init&gt;(Command.java:50)
at hudson.remoting.Request.&lt;init&gt;(Request.java:98)
at hudson.remoting.UserRequest.&lt;init&gt;(UserRequest.java:59)
at hudson.remoting.Channel.callAsync(Channel.java:749)
at hudson.node_monitors.AbstractAsyncNodeMonitorDescriptor.monitor(AbstractAsyncNodeMonitorDescriptor.java:76)
at hudson.node_monitors.ResponseTimeMonitor$1.monitor(ResponseTimeMonitor.java:57)
at hudson.node_monitors.AbstractNodeMonitorDescriptor$Record.run(AbstractNodeMonitorDescriptor.java:280)
Feb 10, 2014 5:24:41 PM hudson.remoting.Channel$2 handle
SEVERE: Failed to execute command UserRequest:hudson.node_monitors.SwapSpaceMonitor$MonitorTask@438a24d1 (channel channel)
java.lang.OutOfMemoryError: unable to create <span class="code-keyword">new</span> <span class="code-keyword">native</span> thread
at java.lang.<span class="code-object">Thread</span>.start0(Native Method)
at java.lang.<span class="code-object">Thread</span>.start(<span class="code-object">Thread</span>.java:597)
at java.util.concurrent.ThreadPoolExecutor.addIfUnderMaximumPoolSize(ThreadPoolExecutor.java:727)
at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:657)
at java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:92)
at hudson.remoting.DelegatingExecutorService.submit(DelegatingExecutorService.java:42)
at hudson.remoting.InterceptingExecutorService.submit(InterceptingExecutorService.java:46)
at hudson.remoting.InterceptingExecutorService.submit(InterceptingExecutorService.java:41)
at hudson.remoting.Request.execute(Request.java:307)
at hudson.remoting.Channel$2.handle(Channel.java:461)
at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:60)
Feb 10, 2014 5:24:41 PM hudson.remoting.Channel$2 handle
SEVERE: This command is created here
Command UserRequest:hudson.node_monitors.SwapSpaceMonitor$MonitorTask@438a24d1 created at
at hudson.remoting.Command.&lt;init&gt;(Command.java:67)
at hudson.remoting.Command.&lt;init&gt;(Command.java:50)
at hudson.remoting.Request.&lt;init&gt;(Request.java:98)
at hudson.remoting.UserRequest.&lt;init&gt;(UserRequest.java:59)
at hudson.remoting.Channel.callAsync(Channel.java:749)
at hudson.node_monitors.AbstractAsyncNodeMonitorDescriptor.monitor(AbstractAsyncNodeMonitorDescriptor.java:76)
at hudson.node_monitors.AbstractNodeMonitorDescriptor$Record.run(AbstractNodeMonitorDescriptor.java:280)
Feb 10, 2014 5:24:41 PM hudson.remoting.Channel$2 handle
SEVERE: Failed to execute command Pipe.Chunk(4,8192) (channel channel)
java.lang.OutOfMemoryError: unable to create <span class="code-keyword">new</span> <span class="code-keyword">native</span> thread
at java.lang.<span class="code-object">Thread</span>.start0(Native Method)
at java.lang.<span class="code-object">Thread</span>.start(<span class="code-object">Thread</span>.java:597)
at java.util.concurrent.ThreadPoolExecutor.addIfUnderMaximumPoolSize(ThreadPoolExecutor.java:727)
at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:657)
at java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:92)
at hudson.remoting.DelegatingExecutorService.submit(DelegatingExecutorService.java:42)
at hudson.remoting.InterceptingExecutorService.submit(InterceptingExecutorService.java:46)
at hudson.remoting.InterceptingExecutorService.submit(InterceptingExecutorService.java:41)
at hudson.remoting.SingleLaneExecutorService.execute(SingleLaneExecutorService.java:104)
at java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:78)
at hudson.remoting.PipeWriter.submit(PipeWriter.java:152)
at hudson.remoting.ProxyOutputStream$Chunk.execute(ProxyOutputStream.java:247)
at hudson.remoting.Channel$2.handle(Channel.java:461)
at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:60)
Feb 10, 2014 5:24:41 PM hudson.remoting.Channel$2 handle
SEVERE: This command is created here
Feb 10, 2014 5:24:41 PM hudson.remoting.Channel$2 handle
SEVERE: Failed to execute command UserRequest:hudson.FilePath$9@480e8c6 (channel channel)
java.lang.OutOfMemoryError: unable to create <span class="code-keyword">new</span> <span class="code-keyword">native</span> thread
at java.lang.<span class="code-object">Thread</span>.start0(Native Method)
at java.lang.<span class="code-object">Thread</span>.start(<span class="code-object">Thread</span>.java:597)
at java.util.concurrent.ThreadPoolExecutor.addIfUnderMaximumPoolSize(ThreadPoolExecutor.java:727)
at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:657)
at java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:92)
at hudson.remoting.DelegatingExecutorService.submit(DelegatingExecutorService.java:42)
at hudson.remoting.InterceptingExecutorService.submit(InterceptingExecutorService.java:46)
at hudson.remoting.InterceptingExecutorService.submit(InterceptingExecutorService.java:41)
at hudson.remoting.Request.execute(Request.java:307)
at hudson.remoting.Channel$2.handle(Channel.java:461)
at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:60)
SEVERE: This command is created here
Command UserRequest:hudson.node_monitors.ResponseTimeMonitor$Step1@76ff540f created at
at hudson.remoting.Command.&lt;init&gt;(Command.java:67)
at hudson.remoting.Command.&lt;init&gt;(Command.java:50)
at hudson.remoting.Request.&lt;init&gt;(Request.java:98)
at hudson.remoting.UserRequest.&lt;init&gt;(UserRequest.java:59)
at hudson.remoting.Channel.callAsync(Channel.java:749)
at hudson.node_monitors.AbstractAsyncNodeMonitorDescriptor.monitor(AbstractAsyncNodeMonitorDescriptor.java:76)
at hudson.node_monitors.ResponseTimeMonitor$1.monitor(ResponseTimeMonitor.java:57)
at hudson.node_monitors.AbstractNodeMonitorDescriptor$Record.run(AbstractNodeMonitorDescriptor.java:280)
Feb 10, 2014 5:25:17 PM hudson.remoting.Channel$2 handle
SEVERE: Failed to execute command UserRequest:hudson.FilePath$9@4a50a2df (channel channel)
java.lang.OutOfMemoryError: unable to create <span class="code-keyword">new</span> <span class="code-keyword">native</span> thread
at java.lang.<span class="code-object">Thread</span>.start0(Native Method)
at java.lang.<span class="code-object">Thread</span>.start(<span class="code-object">Thread</span>.java:597)
at java.util.concurrent.ThreadPoolExecutor.addIfUnderMaximumPoolSize(ThreadPoolExecutor.java:727)
at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:657)
at java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:92)
at hudson.remoting.DelegatingExecutorService.submit(DelegatingExecutorService.java:42)
at hudson.remoting.InterceptingExecutorService.submit(InterceptingExecutorService.java:46)
at hudson.remoting.InterceptingExecutorService.submit(InterceptingExecutorService.java:41)
at hudson.remoting.Request.execute(Request.java:307)
at hudson.remoting.Channel$2.handle(Channel.java:461)
at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:60)
Feb 10, 2014 5:25:17 PM hudson.remoting.Channel$2 handle
SEVERE: This command is created here
Command UserRequest:hudson.FilePath$9@4a50a2df created at
at hudson.remoting.Command.&lt;init&gt;(Command.java:67)
at hudson.remoting.Command.&lt;init&gt;(Command.java:50)
at hudson.remoting.Request.&lt;init&gt;(Request.java:98)
at hudson.remoting.UserRequest.&lt;init&gt;(UserRequest.java:59)
at hudson.remoting.Channel.callAsync(Channel.java:749)
at hudson.node_monitors.AbstractAsyncNodeMonitorDescriptor.monitor(AbstractAsyncNodeMonitorDescriptor.java:76)
at hudson.node_monitors.AbstractNodeMonitorDescriptor$Record.run(AbstractNodeMonitorDescriptor.java:280)
Feb 10, 2014 5:25:18 PM hudson.remoting.Channel$2 handle
SEVERE: Failed to execute command UserRequest:hudson.FilePath$9@77d5fff1 (channel channel)
java.lang.OutOfMemoryError: unable to create <span class="code-keyword">new</span> <span class="code-keyword">native</span> thread
at java.lang.<span class="code-object">Thread</span>.start0(Native Method)
at java.lang.<span class="code-object">Thread</span>.start(<span class="code-object">Thread</span>.java:597)
at java.util.concurrent.ThreadPoolExecutor.addIfUnderMaximumPoolSize(ThreadPoolExecutor.java:727)
at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:657)
at java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:92)
at hudson.remoting.DelegatingExecutorService.submit(DelegatingExecutorService.java:42)
at hudson.remoting.InterceptingExecutorService.submit(InterceptingExecutorService.java:46)
at hudson.remoting.InterceptingExecutorService.submit(InterceptingExecutorService.java:41)
at hudson.remoting.Request.execute(Request.java:307)
at hudson.remoting.Channel$2.handle(Channel.java:461)
at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:60)
Feb 10, 2014 5:25:18 PM hudson.remoting.Channel$2 handle
SEVERE: This command is created here
Command UserRequest:hudson.FilePath$9@77d5fff1 created at
at hudson.remoting.Command.&lt;init&gt;(Command.java:67)
at hudson.remoting.Command.&lt;init&gt;(Command.java:50)
at hudson.remoting.Request.&lt;init&gt;(Request.java:98)
at hudson.remoting.UserRequest.&lt;init&gt;(UserRequest.java:59)
at hudson.remoting.Channel.callAsync(Channel.java:749)
at hudson.node_monitors.AbstractAsyncNodeMonitorDescriptor.monitor(AbstractAsyncNodeMonitorDescriptor.java:76)
at hudson.node_monitors.AbstractNodeMonitorDescriptor$Record.run(AbstractNodeMonitorDescriptor.java:280)
Feb 10, 2014 5:25:18 PM hudson.remoting.Channel$2 handle
SEVERE: Failed to execute command UserRequest:hudson.FilePath$9@3c51eb73 (channel channel)
java.lang.OutOfMemoryError: unable to create <span class="code-keyword">new</span> <span class="code-keyword">native</span> thread
at java.lang.<span class="code-object">Thread</span>.start0(Native Method)
at java.lang.<span class="code-object">Thread</span>.start(<span class="code-object">Thread</span>.java:597)
at java.util.concurrent.ThreadPoolExecutor.addIfUnderMaximumPoolSize(ThreadPoolExecutor.java:727)
at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:657)
at java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:92)
at hudson.remoting.DelegatingExecutorService.submit(DelegatingExecutorService.java:42)
at hudson.remoting.InterceptingExecutorService.submit(InterceptingExecutorService.java:46)
at hudson.remoting.InterceptingExecutorService.submit(InterceptingExecutorService.java:41)
at hudson.remoting.Request.execute(Request.java:307)
at hudson.remoting.Channel$2.handle(Channel.java:461)
at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:60)
Feb 10, 2014 5:25:18 PM hudson.remoting.Channel$2 handle
SEVERE: This command is created here
Command UserRequest:hudson.FilePath$9@3c51eb73 created at
at hudson.remoting.Command.&lt;init&gt;(Command.java:67)
at hudson.remoting.Command.&lt;init&gt;(Command.java:50)
at hudson.remoting.Request.&lt;init&gt;(Request.java:98)
at hudson.remoting.UserRequest.&lt;init&gt;(UserRequest.java:59)
at hudson.remoting.Channel.callAsync(Channel.java:749)
at hudson.node_monitors.AbstractAsyncNodeMonitorDescriptor.monitor(AbstractAsyncNodeMonitorDescriptor.java:76)
at hudson.node_monitors.AbstractNodeMonitorDescriptor$Record.run(AbstractNodeMonitorDescriptor.java:280)
2014-02-10 18:53:15
Full thread dump Java HotSpot(TM) 64-Bit Server VM (16.3-b01 mixed mode):

<span class="code-quote">"pool-1-thread-281"</span> prio=10 tid=0x00007fa9bc001000 nid=0x3150 waiting on condition [0x00007faa15bdc000]
java.lang.<span class="code-object">Thread</span>.State: TIMED_WAITING (parking)
at sun.misc.Unsafe.park(Native Method)
 - parking to wait <span class="code-keyword">for</span> &lt;0x00007facf8c32d60&gt; (a java.util.concurrent.SynchronousQueue$TransferStack)
at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:198)
at java.util.concurrent.SynchronousQueue$TransferStack.awaitFulfill(SynchronousQueue.java:424)
at java.util.concurrent.SynchronousQueue$TransferStack.transfer(SynchronousQueue.java:323)
at java.util.concurrent.SynchronousQueue.poll(SynchronousQueue.java:874)
at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:945)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:907)
at java.lang.<span class="code-object">Thread</span>.run(<span class="code-object">Thread</span>.java:619)

<span class="code-quote">"pool-1-thread-279 <span class="code-keyword">for</span> channel"</span> prio=10 tid=0x00007fa9bc00e800 nid=0x2613 waiting on condition [0x00007faa157d8000]
java.lang.<span class="code-object">Thread</span>.State: WAITING (parking)
at sun.misc.Unsafe.park(Native Method)
 - parking to wait <span class="code-keyword">for</span> &lt;0x00007face7a65b58&gt; (a java.util.concurrent.FutureTask$Sync)
at java.util.concurrent.locks.LockSupport.park(LockSupport.java:158)
at java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(AbstractQueuedSynchronizer.java:747)
at java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireSharedInterruptibly(AbstractQueuedSynchronizer.java:905)
at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireSharedInterruptibly(AbstractQueuedSynchronizer.java:1217)
at java.util.concurrent.FutureTask$Sync.innerGet(FutureTask.java:218)
at java.util.concurrent.FutureTask.get(FutureTask.java:83)
at hudson.remoting.Request$2.run(Request.java:326)
at hudson.remoting.InterceptingExecutorService$1.call(InterceptingExecutorService.java:72)
at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:303)
at java.util.concurrent.FutureTask.run(FutureTask.java:138)
at java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:886)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:908)
at java.lang.<span class="code-object">Thread</span>.run(<span class="code-object">Thread</span>.java:619)

<span class="code-quote">"Ping thread <span class="code-keyword">for</span> channel hudson.remoting.Channel@37bd2664:channel"</span> daemon prio=10 tid=0x00007fa9b4020800 nid=0x4158 waiting on condition [0x00007faa16dec000]
java.lang.<span class="code-object">Thread</span>.State: TIMED_WAITING (sleeping)
at java.lang.<span class="code-object">Thread</span>.sleep(Native Method)
at hudson.remoting.PingThread.run(PingThread.java:86)

<span class="code-quote">"<span class="code-object">Thread</span>-1 / waiting <span class="code-keyword">for</span> hudson.remoting.Channel@37bd2664:channel"</span> daemon prio=10 tid=0x00007fa9b4003000 nid=0x4156 in <span class="code-object">Object</span>.wait() [0x00007faa16fee000]
java.lang.<span class="code-object">Thread</span>.State: TIMED_WAITING (on object monitor)
at java.lang.<span class="code-object">Object</span>.wait(Native Method)
 - waiting on &lt;0x00007facf7f88240&gt; (a hudson.remoting.RemoteInvocationHandler$RPCRequest)
at hudson.remoting.Request.call(Request.java:146)
 - locked &lt;0x00007facf7f88240&gt; (a hudson.remoting.RemoteInvocationHandler$RPCRequest)
at hudson.remoting.RemoteInvocationHandler.invoke(RemoteInvocationHandler.java:170)
at $Proxy4.writeJarTo(Unknown Source)
at hudson.remoting.FileSystemJarCache.retrieve(FileSystemJarCache.java:66)
at hudson.remoting.JarCacheSupport$1.run(JarCacheSupport.java:62)
at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:441)
at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:303)
at java.util.concurrent.FutureTask.run(FutureTask.java:138)
at hudson.remoting.AtmostOneThreadExecutor$Worker.run(AtmostOneThreadExecutor.java:104)
at java.lang.<span class="code-object">Thread</span>.run(<span class="code-object">Thread</span>.java:619)

<span class="code-quote">"Ping thread <span class="code-keyword">for</span> channel hudson.remoting.Channel@37bd2664:channel"</span> daemon prio=10 tid=0x00007fae381ae800 nid=0x4154 waiting on condition [0x00007faa171f0000]
java.lang.<span class="code-object">Thread</span>.State: TIMED_WAITING (sleeping)
at java.lang.<span class="code-object">Thread</span>.sleep(Native Method)
at hudson.remoting.PingThread.run(PingThread.java:86)

<span class="code-quote">"Channel reader thread: channel"</span> prio=10 tid=0x00007fae381bf800 nid=0x4153 runnable [0x00007faa172f1000]
java.lang.<span class="code-object">Thread</span>.State: RUNNABLE
at java.io.FileInputStream.readBytes(Native Method)
at java.io.FileInputStream.read(FileInputStream.java:199)
at java.io.BufferedInputStream.fill(BufferedInputStream.java:218)
at java.io.BufferedInputStream.read(BufferedInputStream.java:237)
 - locked &lt;0x00007facf7f921f8&gt; (a java.io.BufferedInputStream)
at hudson.remoting.FlightRecorderInputStream.read(FlightRecorderInputStream.java:77)
at java.io.ObjectInputStream$PeekInputStream.peek(ObjectInputStream.java:2249)
at java.io.ObjectInputStream$BlockDataInputStream.peek(ObjectInputStream.java:2542)
at java.io.ObjectInputStream$BlockDataInputStream.peekByte(ObjectInputStream.java:2552)
at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1297)
at java.io.ObjectInputStream.readObject(ObjectInputStream.java:351)
at hudson.remoting.Command.readFrom(Command.java:92)
at hudson.remoting.ClassicCommandTransport.read(ClassicCommandTransport.java:71)
at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:48)

<span class="code-quote">"Low Memory Detector"</span> daemon prio=10 tid=0x00007fae380d9800 nid=0x4151 runnable [0x0000000000000000]
java.lang.<span class="code-object">Thread</span>.State: RUNNABLE

<span class="code-quote">"CompilerThread1"</span> daemon prio=10 tid=0x00007fae380d7000 nid=0x4150 waiting on condition [0x0000000000000000]
java.lang.<span class="code-object">Thread</span>.State: RUNNABLE

<span class="code-quote">"CompilerThread0"</span> daemon prio=10 tid=0x00007fae380d4800 nid=0x414f waiting on condition [0x0000000000000000]
java.lang.<span class="code-object">Thread</span>.State: RUNNABLE

<span class="code-quote">"Signal Dispatcher"</span> daemon prio=10 tid=0x00007fae380d2800 nid=0x414e waiting on condition [0x0000000000000000]
java.lang.<span class="code-object">Thread</span>.State: RUNNABLE

<span class="code-quote">"Finalizer"</span> daemon prio=10 tid=0x00007fae380b4800 nid=0x414d in <span class="code-object">Object</span>.wait() [0x00007faa178f7000]
java.lang.<span class="code-object">Thread</span>.State: WAITING (on object monitor)
at java.lang.<span class="code-object">Object</span>.wait(Native Method)
 - waiting on &lt;0x00007facf7f927c0&gt; (a java.lang.ref.ReferenceQueue$Lock)
at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:118)
 - locked &lt;0x00007facf7f927c0&gt; (a java.lang.ref.ReferenceQueue$Lock)
at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:134)
at java.lang.ref.Finalizer$FinalizerThread.run(Finalizer.java:159)

<span class="code-quote">"Reference Handler"</span> daemon prio=10 tid=0x00007fae380b2800 nid=0x414c in <span class="code-object">Object</span>.wait() [0x00007faa179f8000]
java.lang.<span class="code-object">Thread</span>.State: WAITING (on object monitor)
at java.lang.<span class="code-object">Object</span>.wait(Native Method)
 - waiting on &lt;0x00007facf7f8b760&gt; (a java.lang.ref.Reference$Lock)
at java.lang.<span class="code-object">Object</span>.wait(<span class="code-object">Object</span>.java:485)
at java.lang.ref.Reference$ReferenceHandler.run(Reference.java:116)
 - locked &lt;0x00007facf7f8b760&gt; (a java.lang.ref.Reference$Lock)

<span class="code-quote">"main"</span> prio=10 tid=0x00007fae38006000 nid=0x4133 in <span class="code-object">Object</span>.wait() [0x00007fae3d158000]
java.lang.<span class="code-object">Thread</span>.State: WAITING (on object monitor)
at java.lang.<span class="code-object">Object</span>.wait(Native Method)
 - waiting on &lt;0x00007facf7f882a8&gt; (a hudson.remoting.Channel)
at java.lang.<span class="code-object">Object</span>.wait(<span class="code-object">Object</span>.java:485)
at hudson.remoting.Channel.join(Channel.java:850)
 - locked &lt;0x00007facf7f882a8&gt; (a hudson.remoting.Channel)
at hudson.remoting.Launcher.main(Launcher.java:523)
at hudson.remoting.Launcher.runWithStdinStdout(Launcher.java:457)
at hudson.remoting.Launcher.run(Launcher.java:238)
at hudson.remoting.Launcher.main(Launcher.java:192)
</pre>
</div></div>

<p>These errors were not dependably reproducible, and were nearly impossible to duplicate in isolation (i.e. could not reproduce when just running jobs on a single slave node).</p>

<p>After some time debugging and ruling out some things, we determined that perhaps the user nproc limit (ulimit -u) might be too low, based on the errors that w were seeing and what we were able to rule out after debugging. We increased this limit on our slave nodes and...things went back to normal.</p>

<p>None of our build and test jobs changed across the update, meaning that any changes we observed were introduced as a result of the upgrade. My question is then: what has changed recently with slave management that has increased the number of per node threads required? We don't yet have solid watermark data on how high it was actually going, but it was high enough to hit the ceiling for system default (1024).</p></pre>
</details>
</details>
<ul><li><i>environment</i>: <code>CentOS 6</code></li></ul>
